trajopt/src/collision_terms.cpp

```diff
@@ -159,6 +159,225 @@ void DebugPrintInfo(const tesseract_collision::ContactResult& res,
+
+    // Eigen::Isometry3d test_link_transform, temp1, temp2;
+    // manip->calcFwdKin(test_link_transform, dofvals, itA->link_name);
+    // temp1 = world_to_base * test_link_transform;
+    // temp2 = link_transform * itA->transform.inverse();
+    // assert(temp1.isApprox(temp2, 0.0001));
+
+    // Eigen::MatrixXd jac_test;
+    // jac_test.resize(6, manip->numJoints());
+    // tesseract_kinematics::numericalJacobian(jac_test, world_to_base, *manip, dofvals, itA->link_name,
+    // res.nearest_points_local[0]); bool check = jac.isApprox(jac_test, 1e-3); assert(check == true);
+
+    results.gradients[0].gradient = -contact_result.normal.transpose() * jac.topRows(3);
+  }
+
+  tesseract_environment::AdjacencyMapPair::ConstPtr itB = adjacency_map_->getLinkMapping(contact_result.link_names[1]);
+  if (itB != nullptr)
+  {
+    results.gradients[1].has_gradient = true;
+
+    Eigen::MatrixXd jac;
+    jac.resize(6, manip_->numJoints());
+
+    if (contact_result.cc_type[1] == tesseract_collision::ContinuousCollisionType::CCType_Time0)
+      dofvalst = dofvals0;
+    else if (contact_result.cc_type[1] == tesseract_collision::ContinuousCollisionType::CCType_Time1)
+      dofvalst = dofvals0;
+    else
+      dofvalst = dofvals0 + (dofvals1 - dofvals0) * contact_result.cc_time[1];
+
+    // Calculate Jacobian
+    manip_->calcJacobian(jac, dofvalst, itB->link_name);
+
+    // Need to change the base and ref point of the jacobian.
+    // When changing ref point you must provide a vector from the current ref
+    // point to the new ref point.
+    results.gradients[1].scale = 1;
+    Eigen::Isometry3d link_transform = contact_result.transform[1];
+
+    assert(contact_result.cc_time[1] >= 0.0 && contact_result.cc_time[1] <= 1.0);
+    results.gradients[1].scale = (isTimestep1) ? contact_result.cc_time[1] : (1 - contact_result.cc_time[1]);
+    link_transform = (isTimestep1) ? contact_result.cc_transform[1] : contact_result.transform[1];
+
+    tesseract_kinematics::jacobianChangeBase(jac, world_to_base_);
+    tesseract_kinematics::jacobianChangeRefPoint(jac,
+                                                 (link_transform * itB->transform.inverse()).linear() *
+                                                     (itB->transform * contact_result.nearest_points_local[1]));
+
+    // Eigen::Isometry3d test_link_transform, temp1, temp2;
+    // manip->calcFwdKin(test_link_transform, dofvals, itB->link_name);
+    // temp1 = world_to_base * test_link_transform;
+    // temp2 = link_transform * itB->transform.inverse();
+    // assert(temp1.isApprox(temp2, 0.0001));
+
+    // Eigen::MatrixXd jac_test;
+    // jac_test.resize(6, manip->numJoints());
+    // tesseract_kinematics::numericalJacobian(jac_test, world_to_base, *manip, dofvals, itB->link_name,
+    // res.nearest_points_local[1]); bool check = jac.isApprox(jac_test, 1e-3); assert(check == true);
+
+    results.gradients[1].gradient = contact_result.normal.transpose() * jac.topRows(3);
+  }
```

Comment:
Same comment as above about consolidating into a for loop. It also looks like you're duplicating a lot of code from previous functions. Is there a way to reformat this overload to utilize the logic the "main" function rather than copying the content?

Reply:
I added the for loop.